kotlin/iwa-sample-app/src/main/java/com/esri/android/samples/iwasampleapp/MainActivity.kt

```diff
@@ -0,0 +1,173 @@
+ * 2. Tap on the "Sign In" button to connect to the portal as a named user. A dialog will be popped up for
+ * username/password. When a correct credential is passed, portal will be loaded. If you keep passing in
+ * wrong credential, no more prompt for credential after 5 attempts, which is actually 4 times due to a known
+ * issue in our SDK. Then portal will fail to load. You can click the "Cancel" button to cancel the sign-in
+ * process. Portal will be failed to load with different error message.
+ */
+class MainActivity : AppCompatActivity() {
+
+    private lateinit var map: ArcGISMap
+    
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+
+        // create a Streets BaseMap
+        map = ArcGISMap(Basemap.createStreets())
+        // set the map to be displayed in this view
+        mapView.map = map
+        
+        // Set authentication challenge handler
+        AuthenticationManager.setAuthenticationChallengeHandler(IWACustomChallengeHandler(this))
+        
+        // Sign in a portal
+        signinButton.setOnClickListener{
+            // Validate portal url
+            val url : String ? = portalUrl.text.toString()
+            if (url.isNullOrEmpty()) {
+                Toast.makeText(applicationContext, "Portal url is empty. Please enter portal url!", Toast.LENGTH_SHORT).show()
+                return@setOnClickListener
+            }
+            
+            // Clear credential cache. This is for test purpose. In a real app credential cache provides
+            // better user experience.
+            AuthenticationManager.CredentialCache.clear()
+            // Create a Portal object with loginRequired as true, which forces user to sign in
+            val portal : Portal = Portal(url, true)
+            portal.loadAsync()
+            portal.addDoneLoadingListener{
+                if (portal.loadStatus == LoadStatus.LOADED) {
+                    // Portal is loaded. Add more login here
+                    Toast.makeText(applicationContext, "Portal is loaded!", Toast.LENGTH_SHORT).show()
+                } else {
+                    // Portal is failed to load. Handler load errors
+                    val error = portal.loadError
+                    if (error.errorCode == 17) {
+                        // User canceled exception
+                        Toast.makeText(applicationContext, "Portal loading is cancelled by user!", Toast.LENGTH_SHORT).show()
+                    } else {
+                        // Other failures
+                        Toast.makeText(applicationContext, 
+                                "Portal fails to load: " + error.cause?.message ?: error.message, Toast.LENGTH_SHORT).show()
+                    }
+                }
+            }
+        }
+    }
+}
+
+const val MAX_ATTEMPTS = 5
+
+/**
+ * A custom authentication challenge handler to handle user credential challenge.
```

Comment:
plural: "... to handle user credential challenges."

Reply:
updated